File: python/src/dorado/skein.py
"""From-scratch Skein-512 hash and MAC (Skein 1.3), built on Threefish-512 via UBI
(Unique Block Iteration), the hash function Threefish was designed to power. The
output length is chosen per call. Python port of the dorado Rust crate's skein
module. Streams via the incremental Skein512 class; one-shot helpers wrap it.
"""

from __future__ import annotations

import struct
from typing import BinaryIO

from .threefish import Threefish

_BLOCK = 64

# UBI block type values (the 6-bit type field of the tweak).
_T_KEY = 0
_T_CFG = 4
_T_MSG = 48
_T_OUT = 63


def _tweak(position: int, ty: int, first: bool, last: bool) -> bytes:
    t1 = ty << 56
    if first:
        t1 |= 1 << 62
    if last:
        t1 |= 1 << 63
    return struct.pack("<QQ", position, t1)


def _ubi(g: bytearray, msg: bytes, ty: int) -> None:
    total = len(msg)
    offset = 0
    position = 0
    first = True
    while True:
        take = min(total - offset, _BLOCK)
        block = msg[offset : offset + take] + bytes(_BLOCK - take)
        position += take
        offset += take
        last = offset >= total
        enc = Threefish.t512(bytes(g), _tweak(position, ty, first, last)).encrypt_block(block)
        for i in range(_BLOCK):
            g[i] = enc[i] ^ block[i]
        first = False
        if last:
            break


def _config_block(out_bits: int) -> bytes:
    c = bytearray(32)
    c[0:4] = b"SHA3"
    c[4] = 1  # version 1 (16-bit little-endian)
    struct.pack_into("<Q", c, 8, out_bits)
    return bytes(c)


def _output_into(g: bytes, out_len: int) -> bytes:
    out = bytearray()
    counter = 0
    while len(out) < out_len:
        block = bytearray(g)
        _ubi(block, struct.pack("<Q", counter), _T_OUT)
        out += block[: min(_BLOCK, out_len - len(out))]
        counter += 1
    return bytes(out[:out_len])


class Skein512:
    """Incremental Skein-512 hash or MAC, holding back the final block until done."""

    def __init__(self, out_len: int, key: bytes = b""):
        self._out_len = out_len
        g = bytearray(64)
        if key:
            _ubi(g, key, _T_KEY)
        _ubi(g, _config_block(out_len * 8), _T_CFG)
        self._g = g
        self._buffer = bytearray()
        self._position = 0
        self._first = True

    def _commit(self, block: bytes, nbytes: int, last: bool) -> None:
        self._position += nbytes
        enc = Threefish.t512(bytes(self._g), _tweak(self._position, _T_MSG, self._first, last)).encrypt_block(block)
        for i in range(_BLOCK):
            self._g[i] = enc[i] ^ block[i]
        self._first = False

    def update(self, data: bytes) -> "Skein512":
        self._buffer += data
        # Process all but the final (possibly partial) block; only finalize knows last.
        while len(self._buffer) > _BLOCK:
            self._commit(bytes(self._buffer[:_BLOCK]), _BLOCK, False)
            del self._buffer[:_BLOCK]
        return self

    def digest(self) -> bytes:
        g_save = bytearray(self._g)
        pos_save = self._position
        first_save = self._first
        block = bytes(self._buffer) + bytes(_BLOCK - len(self._buffer))
        self._commit(block, len(self._buffer), True)
        out = _output_into(bytes(self._g), self._out_len)
        # Restore so update() may continue after a digest().
        self._g = g_save
        self._position = pos_save
        self._first = first_save
        return out


def hash(out_len: int, msg: bytes) -> bytes:
    """One-shot Skein-512 hash of msg producing out_len bytes."""
    return Skein512(out_len).update(msg).digest()


def mac(key: bytes, out_len: int, msg: bytes) -> bytes:
    """One-shot Skein-512 keyed MAC producing out_len bytes (key absorbed first)."""
    return Skein512(out_len, key).update(msg).digest()


def hash_stream(out_len: int, reader: BinaryIO, buf_size: int = 64 * 1024) -> bytes:
    """Stream-hash a file-like object in constant memory."""
    h = Skein512(out_len)
    while True:
        chunk = reader.read(buf_size)
        if not chunk:
            break
        h.update(chunk)
    return h.digest()

File: haskell/src/Dorado/Skein.hs
{-# LANGUAGE BangPatterns #-}

-- | Skein-512 hash and MAC (Skein 1.3), built on Threefish-512 via UBI (Unique
-- Block Iteration), from scratch. Skein is the hash Threefish was designed to
-- power: it threads message blocks through Threefish-512, using the tweak to
-- encode each block's byte position and type, and xors the plaintext block back
-- in (Matyas-Meyer-Oseas).
--
-- This mirrors the Rust reference's @skein.rs@; verified against the digests the
-- Rust @gyotaku@ tool produces. The output length is fixed per call because
-- Skein folds it into the configuration block that seeds the chaining value.
--
-- These are pure one-shot functions: the chaining value from folding the whole
-- message through one UBI pass is identical to the incremental block-by-block
-- form, so a streaming variant (for hashing inputs larger than memory) can be
-- added later without changing results.
module Dorado.Skein
  ( hash
  , mac
  ) where

import Data.Bits (shiftL, shiftR, xor, (.&.), (.|.))
import Data.Word (Word64)
import qualified Data.ByteString as BS
import Data.ByteString (ByteString)

import Dorado.Threefish (Variant (TF512), encryptBlock, newThreefish)

block :: Int
block = 64

-- UBI block-type values (the 6-bit type field of the tweak).
tKey, tCfg, tMsg, tOut :: Word64
tKey = 0
tCfg = 4
tMsg = 48
tOut = 63

-- | Eight little-endian bytes of a 'Word64'.
le64 :: Word64 -> ByteString
le64 w = BS.pack [ fromIntegral (w `shiftR` (8 * k)) .&. 0xff | k <- [0 .. 7] ]

-- | The 128-bit UBI tweak (16 little-endian bytes) for a block at byte
-- @position@ of a pass of type @ty@, with the first/final flags. Layout: bits
-- 0-95 position (only the low 64 used), bits 120-125 type, bit 126 first, bit
-- 127 final.
tweak :: Word64 -> Word64 -> Bool -> Bool -> ByteString
tweak position ty first final = le64 position <> le64 t1
  where
    t1 = (ty `shiftL` 56)
         .|. (if first then 1 `shiftL` 62 else 0)
         .|. (if final then 1 `shiftL` 63 else 0)

-- | One UBI pass: chain @msg@ into the 64-byte chaining value @g@ under block
-- type @ty@. An empty message processes a single zero block at position 0.
ubi :: ByteString -> ByteString -> Word64 -> ByteString
ubi g0 msg ty = go g0 0 0 True
  where
    total = BS.length msg
    go !g !offset !position !first =
      let take' = min (total - offset) block
          raw = BS.take take' (BS.drop offset msg)
          blk = raw <> BS.replicate (block - take') 0
          position' = position + fromIntegral take'
          offset' = offset + take'
          final = offset' >= total
          cipher = newThreefish TF512 g (tweak position' ty first final)
          enc = encryptBlock cipher blk
          g' = BS.pack (BS.zipWith xor enc blk)
      in if final then g' else go g' offset' position' False

-- | The 32-byte Skein configuration block for an output of @outBits@ bits:
-- "SHA3" schema id, version 1, then the output length.
configBlock :: Word64 -> ByteString
configBlock outBits =
  BS.pack [0x53, 0x48, 0x41, 0x33, 1, 0, 0, 0] <> le64 outBits <> BS.replicate 16 0

-- | Produce @outLen@ output bytes from the final chaining value by running the
-- output UBI over an incrementing counter. Generates exactly as many blocks as
-- needed (no lazy-infinite-list reliance).
output :: ByteString -> Int -> ByteString
output g outLen = BS.take outLen (BS.concat [ blockFor c | c <- [0 .. nBlocks - 1] ])
  where
    nBlocks = fromIntegral ((outLen + block - 1) `div` block) :: Word64
    blockFor counter = ubi g (le64 counter) tOut

-- | Skein-512 hash of @msg@ producing @outLen@ bytes.
hash :: Int -> ByteString -> ByteString
hash outLen msg = output g1 outLen
  where
    g0 = ubi (BS.replicate block 0) (configBlock (fromIntegral outLen * 8)) tCfg
    g1 = ubi g0 msg tMsg

-- | Skein-512 MAC (keyed hash) producing @outLen@ bytes; absorbs @key@ through a
-- Key UBI first. An empty key is identical to 'hash'.
mac :: ByteString -> Int -> ByteString -> ByteString
mac key outLen msg = output g1 outLen
  where
    gKey | BS.null key = BS.replicate block 0
         | otherwise   = ubi (BS.replicate block 0) key tKey
    g0 = ubi gKey (configBlock (fromIntegral outLen * 8)) tCfg
    g1 = ubi g0 msg tMsg
